eth/db/chain.py

```diff
@@ -75,15 +86,92 @@ class ChainDB(HeaderDB, ChainDatabaseAPI):
     def __init__(self, db: AtomicDatabaseAPI) -> None:
         self.db = db
 
+    def get_chain_gaps(self) -> ChainGaps:
+        return self._get_chain_gaps(self.db)
+
+    @classmethod
+    def _get_chain_gaps(cls, db: DatabaseAPI) -> ChainGaps:
+        try:
+            encoded_gaps = db[SchemaV1.make_chain_gaps_lookup_key()]
+        except KeyError:
+            return GENESIS_CHAIN_GAPS
+        else:
+            return rlp.decode(encoded_gaps, sedes=chain_gaps)
+
+    @classmethod
+    def _update_chain_gaps(
+            cls,
+            db: DatabaseAPI,
+            persisted_block: BlockAPI,
+            base_gaps: ChainGaps = None
+    ) -> GapInfo:
+
+        # If we make many updates in a row, we can avoid reloading the integrity info by
+        # continuously caching it and providing it as a parameter to this API
+        if base_gaps is None:
+            base_gaps = cls._get_chain_gaps(db)
+
+        gap_change, gaps = fill_gap(persisted_block.header.block_number, base_gaps)
+        if gap_change is not GapChange.NoChange:
+            db.set(
+                SchemaV1.make_chain_gaps_lookup_key(),
+                rlp.encode(gaps, sedes=chain_gaps)
+            )
+
+        return gap_change, gaps
+
+    @classmethod
+    def _update_header_chain_gaps(
+            cls,
+            db: DatabaseAPI,
+            persisted_header: BlockHeaderAPI,
+            base_gaps: ChainGaps = None
+    ) -> GapInfo:
+        # The only reason we overwrite this here is to be able to detect when the HeaderDB
+        # de-canonicalizes an uncle that should cause re-opening a block gap.
+        gap_change, gaps = super()._update_header_chain_gaps(db, persisted_header, base_gaps)
+        # We have overwritten an existing header, this might be a sign of a de-canonicalized uncle
+        if gap_change is GapChange.NoChange:
+            current_gaps = cls._get_chain_gaps(db)
+            if not is_block_number_in_gap(persisted_header.block_number, current_gaps):
+                # ChainDB believes we have that block, find out if we actually have it.
+                try:
```

Comment:
Not sure if there's any easier way to do this here in the `ChainDB`?  If not then it might even make sense to extract this into a `def block_exists(block_hash`) method?

Reply:
I can be on board with that but I do want to bring up one aspect that might be of importance. 

Currently, we hook into `GapChange.NoChange` and then check if we can load the block to figure out if we need to re-open the gap. If we would eliminate that check and simply re-open the block on every `GapChange.NoChange` then this would mean we would always mark blocks as missing whenever we overwrite headers. I guess this could lead to noticeable more request load. E.g. when we restart Trinity and the gap fill sync starts it will jump back a couple of thousand headers and every overwritten header would re-open a block gap which would result into us unnecessarily re-requesting blocks that we already have.

That's why I thought it would be good to do a little extra effort to figure out if we need to mark the block as gap or not. If you think that isn't a big issue then I'm happy to get rid of that extra check.

Reply:
> I guess this could lead to noticeable more request load. E.g. when we restart Trinity and the gap fill sync starts it will jump back a couple of thousand headers and every overwritten header would re-open a block gap which would result into us unnecessarily re-requesting blocks that we already have.

Ah, my mental model is that the block gap filler would check the local DB to see if the block can be loaded before it asks peers. So it would mean an increase in database hits, which we can probably optimize later.

Before looking into it too deeply, I wonder if we can hook into somewhere else that does this check, since we're interested in a really specific subset of cases, where the block numbers of the old canonical headers match the block numbers of the new canonical headers. So maybe the check can be somewhere that that information is available.

> That's why I thought it would be good to do a little extra effort to figure out if we need to mark the block as gap or not. If you think that isn't a big issue then I'm happy to get rid of that extra check.

Yeah, I definitely get the motivation, and I wouldn't be surprised if we had to come back to it later to speed it up a bit. I just think the balance of preferring the simpler & more general solution is worth the cost of the performance hit for now. :+1:

Reply:
>Before looking into it too deeply, I wonder if we can hook into somewhere else that does this check, since we're interested in a really specific subset of cases, where the block numbers of the old canonical headers match the block numbers of the new canonical headers.

Yes, in fact, we can do that check here. We can load the old header at that block number and compare. The downside of this is that it relies on the fact that the header behind the block number wasn't changed at that instant (it will happen later in the same atomic db write, just not yet). So, if things inside `persist_header_chain` would be refactored, this could fall apart. I personally think that is fine though, especially now that we have tests for it. I also added a test case that demonstrates that simply overwriting the exact same header won't cause us to re-open a block gap.

How do you feel about this solution?